### PR TITLE
1st draft for testing API

### PR DESCRIPTION
**Pull request by nicosomb**

### CODE REVIEW

**Comment by nicosomb** [comment]
here is my 1st test, need refactor, I know it: https://github.com/wallabag/wallabag/blob/v2-api-tests/src/Wallabag/CoreBundle/Tests/Controller/WallabagRestControllerTest.php

**Comment by j0k3r** [comment]
I think you need to rebase it against the `v2` branch

**Comment by nicosomb** [comment]
Merge done. 

PR still in WIP.

**Comment by j0k3r** [comment]
Well, for readability, can you separate the commit where you remove the doc from this PR ? The diff is inreadable :confused: 

Also, you can remove the doc directly from the v2 branch, you don't need to create a PR for that.

File: src/Wallabag/CoreBundle/Controller/WallabagRestController.php [hunk 1]
```diff
@@ -33,7 +33,7 @@ public function getSaltAction($username)
             throw $this->createNotFoundException();
         }
 
-        return $user->getSalt();
+        return array($user->getSalt());
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Controller/WallabagRestController.php]: You need to update the phpDoc then

File: src/Wallabag/CoreBundle/Tests/Controller/WallabagRestControllerTest.php [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+            'HTTP_AUTHORIZATION' => 'Authorization profile="UsernameToken"',
+            'HTTP_x-wsse' => 'X-WSSE: UsernameToken Username="'.$username.'", PasswordDigest="'.$digest.'", Nonce="'.$nonce.'", Created="'.$created.'"',
+        );
+    }
+
+    public function testGetSalt()
+    {
+        $client = $this->createClient();
+        $client->request('GET', '/api/salts/admin.json');
+        $this->assertEquals(200, $client->getResponse()->getStatusCode());
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Tests/Controller/WallabagRestControllerTest.php]: Could you check that the decoded json response is a non empty array?

File: src/Wallabag/CoreBundle/Tests/Controller/WallabagRestControllerTest.php [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    public function testGetEntries()
+    {
+        $client = $this->createClient();
+        $client->request('GET', '/api/salts/admin.json');
+        $salt = json_decode($client->getResponse()->getContent());
+
+        $headers = $this->generateHeaders('admin', 'test', $salt[0]);
+
+        $client->request('GET', '/api/entries', array(), array(), $headers);
+        $this->assertContains('Mailjet', $client->getResponse()->getContent());
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Tests/Controller/WallabagRestControllerTest.php]: You can check that the json response contains more than one element (entry)

**Comment by j0k3r** [comment]
Could you add some tests where headers are completely fucked up?

Like :

```php
'PHP_AUTH_USER' => 'little_poney',
'HTTP_AUTHORIZATION' => 'Authorization profile="NothinRelatedToken"',
'HTTP_x-wsse' => 'X-WSSE: UsernameToken Username="'.$username.'", HackingDigest="'.$digest.'", Nonce="'.$nonce.'", Created="'.$created.'"',
```

Etc ..

File: src/Wallabag/CoreBundle/Tests/Controller/WallabagRestControllerTest.php (hunk 2 above)
Comment (nicosomb) [inline src/Wallabag/CoreBundle/Tests/Controller/WallabagRestControllerTest.php]: OK. But what do I need to return in WallabagRestController::getSaltAction()? `$user->getSalt()?array($user->getSalt()):array();`?

Comment (j0k3r) [inline src/Wallabag/CoreBundle/Tests/Controller/WallabagRestControllerTest.php]: The `getSaltAction` is ok.
Here you need to test `$this->assertNotEmpty(json_decode($client->getResponse()->getContent(), true))`

**Comment by nicosomb** [comment]
I added a test with bad headers. 

I have to change my test for this comment : 
> You can check that the json response contains more than one element (entry)

**Comment by nicosomb** [comment]
ready for review.

File: src/Wallabag/CoreBundle/Tests/Controller/WallabagRestControllerTest.php [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        $client->request('GET', '/api/salts/admin.json');
+        $salt = json_decode($client->getResponse()->getContent());
+
+        $headers = $this->generateHeaders('admin', 'test', $salt[0]);
+
+        $client->request('GET', '/api/entries', array(), array(), $headers);
+
+        $this->assertEquals(200, $client->getResponse()->getStatusCode());
+
+  /*      var_dump(count(json_decode($client->getResponse()->getContent())));
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Tests/Controller/WallabagRestControllerTest.php]: Maybe you can remove this debug :)

**Pull request merged**